lld/MachO/Driver.cpp

```diff
@@ -282,11 +284,83 @@ static void saveThinArchiveToRepro(ArchiveFile const *file) {
           ": Archive::children failed: " + toString(std::move(e)));
 }
 
-static InputFile *addFile(StringRef path, LoadType loadType,
-                          bool isLazy = false, bool isExplicit = true,
-                          bool isBundleLoader = false,
-                          bool isForceHidden = false) {
-  std::optional<MemoryBufferRef> buffer = readFile(path);
+class DeferredFile {
+public:
+  DeferredFile(StringRef path, bool isLazy, MemoryBufferRef buffer)
+      : path(path), isLazy(isLazy), buffer(buffer) {}
+  StringRef path;
+  bool isLazy;
+  MemoryBufferRef buffer;
+};
+using DeferredFiles = std::vector<DeferredFile>;
+
+// Most input files have been mapped but not yet paged in.
+// This code forces the page-ins on multiple threads so
+// the process is not stalled waiting on disk buffer i/o.
+void multiThreadedPageInBackground(const DeferredFiles &deferred) {
+  static const size_t pageSize = Process::getPageSizeEstimate();
+  static size_t totalBytes = 0;
+  std::atomic_int index = 0;
+
+  parallelFor(0, config->readThreads, [&](size_t I) {
```

Comment:
Because of how you are using `parallelFor`, `config->readThreads` is only the maximum number of threads that might be spawn for this, not the exact number of threads, which can be lower, governed by the `-threads` parameter. `parallelFor` uses `llvm::parallel::strategy` internally to decide the actual number of threads, which is setup globally when the driver finds the `-threads` argument.

Using `parallelFor` as its authors intended will also avoid the need of the `index` variable and keeping track of it. I provided a snippet before of how I would switch the strategy to one that fits your idea of having a different number of threads for reading files. It should be in some old comment.

Reply:
See my other comment. I'd missed yours. I'm only following the benchmarks as this is the thrust of this PR.

Reply:
Where are we on this PR? I've removed my benchmarking statements for a release build and looked for pre-existing abstraction for queuing work and found [BackgroundQueue](https://clang.llvm.org/extra//doxygen/classclang_1_1clangd_1_1BackgroundQueue.html) which looked promising but it's part of clangd. What are the blockers now?

Reply:
I get that paralellFor may not use exactly readThreads threads the limit in practice being the number of CPUs on the host which is the default value for the -threads option. My point is it doesn't actually matter the precise number of threads as long as there are more than one tickling input flies into memory. The code using paralellFor at least I understand and it is performant.

Reply:
If `readThreads` does not longer represent the exact number of threads that is used to read, maybe it deserve other name and different argument name and help text for the argument. If this is going to use the `-threads` value anyway, why one needs anything more than `-parallalelize-input-preload` boolean flag, and use `deferred.size()` as the value. And if one is going to use `deferred.size()`, why not use `parallelFor` iterating over `deferred` and remove a bunch of code to handle the indices manually?

PD: it can be less performant that this version, but I will argue that I would prefer less performance and easier code to maintain.

Reply:
Thanks very much for your code, it helps me read your mind. I've pushed another commit which also renames the option to `--read-workers instead` of `--read-threads`. See what you think!